airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog.py

```diff
@@ -0,0 +1,161 @@
+    overrode_metadata_data[id_field] = overrode_metadata_data["definitionId"]
+    del overrode_metadata_data["definitionId"]
+
+    # add in useless fields that are currently required for porting to the actor definition spec
+    overrode_metadata_data["tombstone"] = False
+    overrode_metadata_data["custom"] = False
+    overrode_metadata_data["public"] = True
+
+    # if there is no releaseStage, set it to "alpha"
+    # Note: this is something our current cloud catalog generator does
+    # Note: We will not once this is live
+    if not overrode_metadata_data.get("releaseStage"):
+        overrode_metadata_data["releaseStage"] = "alpha"
+
+    return overrode_metadata_data
+
+
+def is_metadata_catalog_enabled(metadata_definition: dict, catalog_name: str) -> bool:
+    return metadata_definition["data"]["catalogs"][catalog_name]["enabled"]
+
+
+def is_metadata_connector_type(metadata_definition: dict, connector_type: str) -> bool:
+    return metadata_definition["data"]["connectorType"] == connector_type
+
+
+def construct_catalog_from_metadata(catalog_derived_metadata_definitions: List[dict], catalog_name: str) -> dict:
+    catalog_sources = [
+        metadata_to_catalog_entry(metadata, "source", catalog_name)
+        for metadata in catalog_derived_metadata_definitions
+        if is_metadata_catalog_enabled(metadata, catalog_name) and is_metadata_connector_type(metadata, "source")
+    ]
+    catalog_destinations = [
+        metadata_to_catalog_entry(metadata, "destination", catalog_name)
+        for metadata in catalog_derived_metadata_definitions
+        if is_metadata_catalog_enabled(metadata, catalog_name) and is_metadata_connector_type(metadata, "destination")
+    ]
+
+    catalog = {"sources": catalog_sources, "destinations": catalog_destinations}
+
+    return catalog
+
+
+# ASSETS
+
+
+@asset(group_name=GROUP_NAME)
+def cloud_catalog_from_metadata(catalog_derived_metadata_definitions: List[PartialMetadataDefinition]) -> dict:
+    """
+    This asset is used to generate the cloud catalog from the metadata definitions.
+
+    TODO (ben): This asset should be updated to use the GCS metadata definitions once available.
+    """
+    return construct_catalog_from_metadata(catalog_derived_metadata_definitions, "cloud")
+
+
+@asset(group_name=GROUP_NAME)
+def oss_catalog_from_metadata(catalog_derived_metadata_definitions: List[PartialMetadataDefinition]) -> dict:
+    """
+    This asset is used to generate the oss catalog from the metadata definitions.
+
+    TODO (ben): This asset should be updated to use the GCS metadata definitions once available.
+    """
+    return construct_catalog_from_metadata(catalog_derived_metadata_definitions, "oss")
+
+
+@asset(group_name=GROUP_NAME)
+def cloud_sources_dataframe(latest_cloud_catalog_dict: dict) -> OutputDataFrame:
+    sources = latest_cloud_catalog_dict["sources"]
+    return output_dataframe(pd.DataFrame(sources))
```

Comment:
nit: I think a custom decorator can be created and use  as you consistently return the same types in the same way on each assets below.

Reply:
I think in this case, because we're already under the weight of dagster decorators I would like to keep whats happening for each asset much more explicit. 

My worry is two layers of decorators, is two layers of indirection, which can make debugging and understanding a little more difficult.

Reply:
Fair enough 😄